### PR TITLE
[API] reseting password with validation

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         |master 
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

**Comment by arti0090** [comment]
I have intentionally left comments with some changes (that tested on apiplatform v2.6) that will work when endpoint will be PATCH /request-reset-password/{token}  At the moment it does not work. We first need support for that version. There should be also no issue with removing these comments as my last commit is pushed https://github.com/Sylius/Sylius/pull/12391/commits/2149e879c5ac8011801c3cf829c012bfdc33b0b5

File: src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Command;
+
+/** @experimental */
+class ResetPassword implements ResetPasswordTokenAwareInterface
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php:17]: Is `ResetPasswordTokenAwareInterface` necessary? why you can't add it in the command body?
IIRC this token you get from email, so can you explain to me why you choose this way to provide `token` to command?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml [hunk 2]
```diff
@@ -78,6 +78,7 @@
         </itemOperations>
 
         <property name="id" identifier="true" writable="false" />
+        <property name="resetPasswordToken" readable="false" />
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml:81]: do we need to expose this data in response? customer should get this data from email (I guess)

File: src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php (hunk 1 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php:17]: WIll be removed in next iteration

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml (hunk 2 above)
Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Customer.xml:81]: unnecessary change, as at first iteration it called resource of customer. I will remove it now

File: src/Sylius/Behat/Context/Api/Shop/LoginContext.php [hunk 3]
```diff
@@ -18,6 +18,9 @@
 use Sylius\Behat\Client\ApiSecurityClientInterface;
 use Sylius\Behat\Client\Request;
 use Sylius\Behat\Client\ResponseCheckerInterface;
+use Sylius\Behat\Service\SharedStorageInterface;
+use Sylius\Component\User\Model\UserInterface;
+use Symfony\Component\HttpFoundation\Request as HTTPRequest;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/LoginContext.php:23]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+use Sylius\Bundle\ApiBundle\Command\ResetPassword;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\Resource\Metadata\MetadataInterface;
+use Sylius\Component\User\Repository\UserRepositoryInterface;
+use Sylius\Component\User\Security\PasswordUpdaterInterface;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+class ResetPasswordHandler implements MessageHandlerInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php:25]:
```suggestion
final class ResetPasswordHandler implements MessageHandlerInterface
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    public function __invoke(ResetPassword $command): void
+    {
+        /** @var ShopUserInterface|null $user */
+        $user = $this->userRepository->findOneBy(['passwordResetToken' => $command->getResetPasswordToken()]);
+
+        Assert::notNull($user);
+
+        $resetting = $this->metadata->getParameter('resetting');
+        $lifetime = new \DateInterval($resetting['token']['ttl']);
+        if (!$user->isPasswordRequestNonExpired($lifetime)) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php:55]:
```suggestion

        if (!$user->isPasswordRequestNonExpired($lifetime)) {
```

File: src/Sylius/Bundle/ApiBundle/DataProvider/ResetPasswordCustomerDataProvider.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Sylius\Bundle\ApiBundle\DataProvider;
+
+use ApiPlatform\Core\DataProvider\ItemDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use Sylius\Bundle\ApiBundle\Command\ResetPassword;
+use Sylius\Component\User\Repository\UserRepositoryInterface;
+
+/** @experimental */
+final class ResetPasswordCustomerDataProvider implements RestrictedDataProviderInterface, ItemDataProviderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ResetPasswordCustomerDataProvider.php:22]: Probably won't be needed

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RequestResetPasswordTokenHandlerSpec.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+use Sylius\Bundle\ApiBundle\Command\SendResetPasswordEmail;
+use Sylius\Component\Core\Model\ShopUserInterface;
+use Sylius\Component\User\Repository\UserRepositoryInterface;
+use Sylius\Component\User\Security\Generator\GeneratorInterface;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Messenger\Stamp\DispatchAfterCurrentBusStamp;
+
+class RequestResetPasswordTokenHandlerSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/RequestResetPasswordTokenHandlerSpec.php:19]:
```suggestion
final class RequestResetPasswordTokenHandlerSpec extends ObjectBehavior
```

File: src/Sylius/Behat/Context/Api/Shop/LoginContext.php (hunk 3 above)
Comment (arti0090) [reply src/Sylius/Behat/Context/Api/Shop/LoginContext.php:21]: Well it is needed now to create request

File: src/Sylius/Bundle/ApiBundle/DataProvider/ResetPasswordItemDataProvider.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace Sylius\Bundle\ApiBundle\DataProvider;
+
+use ApiPlatform\Core\DataProvider\ItemDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use Sylius\Bundle\ApiBundle\Command\ResetPassword;
+use Sylius\Component\User\Repository\UserRepositoryInterface;
+
+/** @experimental */
+final class ResetPasswordItemDataProvider implements RestrictedDataProviderInterface, ItemDataProviderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataProvider/ResetPasswordItemDataProvider.php:22]: Should be tested with PHPSpec

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                    <attribute name="summary">Password reset</attribute>
+                </attribute>
+            </itemOperation>
+        </itemOperations>
+
+    <property name="resetPasswordToken" identifier="true" />
+    <property name="newPassword" required="true" />
+    <property name="confirmNewPassword" required="true" />
+
+    </resource>
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml:59]:
```suggestion
        <property name="resetPasswordToken" identifier="true" />
        <property name="newPassword" required="true" />
        <property name="confirmNewPassword" required="true" />
    </resource>
```

**Review by AdamKasp — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Validator/Constraints/ConfirmResetPasswordValidator.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use Sylius\Bundle\ApiBundle\Command\ResetPassword;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ConfirmResetPasswordValidator extends ConstraintValidator
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Validator/Constraints/ConfirmResetPasswordValidator.php:22]: Missing spec for this class

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/ResetPasswordHandlerSpec.php [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    function it_resets_password(
+        UserRepositoryInterface $userRepository,
+        ShopUserInterface $shopUser,
+        MetadataInterface $metadata,
+        PasswordUpdaterInterface $passwordUpdater
+    ): void {
+        $userRepository->findOneBy(['passwordResetToken' => 'TOKEN'])->willReturn($shopUser);
+        $metadata->getParameter('resetting')->willReturn(['token' => ['ttl' => 'P5D']]);
+
+        $shopUser->isPasswordRequestNonExpired(Argument::that(function(\DateInterval $dateInterval ) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/ResetPasswordHandlerSpec.php:49]:
```suggestion
        $shopUser->isPasswordRequestNonExpired(Argument::that(function(\DateInterval $dateInterval) {
```
Same below

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/LoginContext.php [hunk 12]
```diff
@@ -79,6 +75,17 @@ public function iResetPasswordForEmailInLocale(string $email, string $localeCode
         $this->iResetIt();
     }
 
+    /**
+     * @When /^I follow link on (my) email to reset my password$/
+     */
+    public function iFollowLinkOnMyEmailToResetPassword(UserInterface $user): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/LoginContext.php:81]:
```suggestion
    public function iFollowLinkOnMyEmailToResetPassword(ShopUserInterface $user): void
```
?

File: src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+    public function getResetPasswordToken(): string
+    {
+        return $this->resetPasswordToken;
+    }
+
+    public function setResetPasswordToken(string $token): void
+    {
+        $this->resetPasswordToken = $token;
+    }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Command/ResetPassword.php:41]: Are these methods needed as `$resetPasswordToken` has public access?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+        $resetting = $this->metadata->getParameter('resetting');
+        $lifetime = new \DateInterval($resetting['token']['ttl']);
+
+        if (!$user->isPasswordRequestNonExpired($lifetime)) {
+            throw new \InvalidArgumentException('Password reset token has expired');
+        }
+
+        if ($command->getResetPasswordToken() !== $user->getPasswordResetToken()) {
+            throw new \InvalidArgumentException('Password reset token do not match.');
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php:61]:
```suggestion
            throw new \InvalidArgumentException('Password reset token does not match.');
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        }
+
+        $user->setPlainPassword($command->newPassword);
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/ResetPasswordHandler.php:64]: Do you check somewhere the new password with the password confirmation?

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="https://api-platform.com/schema/metadata https://api-platform.com/schema/metadata/metadata-2.0.xsd"
+>
+    <resource class="Sylius\Bundle\ApiBundle\Command\ResetPassword" shortName="ResetPassword">
+        <attribute name="route_prefix">shop</attribute>
+        <attribute name="validation_groups">sylius</attribute>
+        <attribute name="messenger">input</attribute>
+
+        <collectionOperations>
+            <collectionOperation name="shop_password_reset_request">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml:24]:
```suggestion
            <collectionOperation name="shop_request_reset_password">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+                    <attribute name="groups">shop:reset_password:create</attribute>
+                </attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">Request password reset</attribute>
+                </attribute>
+            </collectionOperation>
+        </collectionOperations>
+
+        <itemOperations>
+            <itemOperation name="shop_password_reset">
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml:40]:
```suggestion
            <itemOperation name="shop_reset_password">
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+                <attribute name="method">PATCH</attribute>
+                <attribute name="path">/reset-password/{id}</attribute>
+                <attribute name="input">Sylius\Bundle\ApiBundle\Command\ResetPassword</attribute>
+                <attribute name="output">false</attribute>
+                <attribute name="status">202</attribute>
+                <attribute name="denormalization_context">
+                    <attribute name="groups">shop:reset_password:update</attribute>
+                </attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">Password reset</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml:50]:
```suggestion
                    <attribute name="summary">Resets password</attribute>
```
to be consistent with other endpoints

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+                <attribute name="method">POST</attribute>
+                <attribute name="path">/request-reset-password</attribute>
+                <attribute name="input">Sylius\Bundle\ApiBundle\Command\RequestResetPasswordToken</attribute>
+                <attribute name="output">false</attribute>
+                <attribute name="status">202</attribute>
+                <attribute name="denormalization_context">
+                    <attribute name="groups">shop:reset_password:create</attribute>
+                </attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">Request password reset</attribute>
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ResetPassword.xml:34]:
```suggestion
                    <attribute name="summary">Requests password reset</attribute>
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, @arti0090! :tada:

**Pull request merged**